File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());

        for(int i=0; i<nums.size(); i+=1){
            if (i != 0 && nums[i] == nums[i-1]) continue;
            int l = i+1, r = nums.size()-1;
            while (l < r){
                int sum = nums[i] + nums[l] + nums[r];
                if (sum == 0) {
                    ans.push_back({nums[i], nums[l], nums[r]});

                    l+=1;
                    r-=1;

                    while (l<r && nums[l] == nums[l-1]) l+=1;
                    while (l<r && nums[r] == nums[r+1]) r-=1;
                } else if (sum > 0) {
                    r-=1;
                } else {
                    l+=1;
                }
            }
        }

        return ans;
    }
};

File: 15-3sum/3sum.go
func threeSum(nums []int) [][]int {
    sort.Ints(nums)
    var triplets [][]int
    for i := range nums {
        if i != 0 && nums[i] == nums[i-1] {
            continue
        }

        l, r := i + 1, len(nums)-1
        for l < r {
            sum := nums[i] + nums[l] + nums[r]

            if sum == 0 { // We found a triplet
                triplets = append(triplets, []int{nums[i], nums[l], nums[r]})


                l++
                r--

                for l < r && nums[l] == nums[l-1] {
                    l++
                }
                for l < r && nums[r] == nums[r+1] {
                    r--
                }
            } else if sum > 0 {
                r--
            } else {
                l++
            }
        }
    }

    return triplets
}